acceptance-tests/master_cli_test.go

```diff
@@ -0,0 +1,71 @@
+package acceptance_tests
+
+import (
+	"bytes"
+	"fmt"
+	"net"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("Master CLI", func() {
+	It("Works when enabled", func() {
+		opsfileMasterCLI := `---
+- type: replace
+  path: /instance_groups/name=haproxy/jobs/name=haproxy/properties/ha_proxy/master_cli_enable?
+  value: true
+- type: replace
+  path: /instance_groups/name=haproxy/jobs/name=haproxy/properties/ha_proxy/master_cli_bind?
+  value: '0.0.0.0:9001'
+`
+		haproxyInfo, _ := deployHAProxy(baseManifestVars{
+			haproxyBackendPort:    12000,
+			haproxyBackendServers: []string{"127.0.0.1"},
+			deploymentName:        defaultDeploymentName,
+		}, []string{opsfileMasterCLI}, map[string]interface{}{}, true)
+
+		By("The master CLI 'show proc' command works")
+		output, err := haproxyCLICommand("show proc", fmt.Sprintf("%s:9001", haproxyInfo.PublicIP))
+		Expect(err).NotTo(HaveOccurred())
+
+		// Example expected output
+		// #<PID>          <type>          <relative PID>  <reloads>       <uptime>        <version>
+		// 15              master          0               0               0d00h11m09s     2.2.14-a07ac36
+		// # workers
+		// 17              worker          1               0               0d00h11m09s     2.2.14-a07ac36
+		Expect(string(output)).To(MatchRegexp("[0-9]+[\\s]+master"))
+		Expect(string(output)).To(MatchRegexp("[0-9]+[\\s]+worker"))
+	})
+})
+
+func haproxyCLICommand(command string, addr string) (string, error) {
+	conn, err := net.DialTimeout("tcp", addr, 5*time.Second)
+	if err != nil {
+		return "", err
+	}
+	defer conn.Close()
+
+	// HAProxy master CLI does not close connection after writing response,
+	// so we will close after one second
+	if err := conn.SetDeadline(time.Now().Add(time.Second)); err != nil {
```

Comment:
shouldn't the `conn.Close()` above do this?

Reply:
conn.Close never gets called without the deadline. If there is no deadline, then `buffer.ReadFrom(conn)` blocks forever because the HAProxy master CLI never sends EOF after responding to the command

Reply:
A possible cleaner alternative would be to use https://golang.org/pkg/bufio/ to read the first 5 lines then close the connection after we get 5 since we know the response happens to be 5 lines long. However we'd still need the timeout anyway, because if for some reason we don't get 5 lines we don't want to wait forever.

Reply:
ok got it